src/backends/mod.rs

```diff
@@ -35,10 +35,205 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
+use errors::HWTracerError;
+use Tracer;
+use backends::dummy::DummyTracer;
+
 #[cfg(perf_pt)]
 pub mod perf_pt;
 #[cfg(perf_pt)]
 pub use self::perf_pt::PerfPTTracer;
 
+#[derive(Debug, PartialEq)]
+pub enum BackendKind {
+    Dummy,
+    PerfPT,
+}
+
+/// Generic configuration interface for all backends.
+/// A `None` field indicates a default value and attributes which don't apply to a given backend
+/// are checked by the `TracerBuilder` when `build()` is called.
+#[derive(Debug)]
+pub struct BackendConfig {
+    perf_aux_bufsize: Option<usize>,
+    perf_data_bufsize: Option<usize>,
+    new_trace_bufsize: Option<usize>,
+}
+
+/// A builder interface for instantiating backends.
+pub struct TracerBuilder {
+    user_backend: Option<BackendKind>,  // User-requested backend choice.
+    auto_backend: BackendKind,          // Auto-chosen backend if `user_backend` is `None`.
```

Comment:
This attribute seems pointless? Surely it should be the job of some later phase to say "if user_backend is None: self.default_platform_backend()"?

Reply:
We have to populate it early so that the user can set a custom configuration on an auto-chosen backend. See "Make thread tracers for a automatically chosen default backend, but with a custom config if PerfPT is chosen" above.